contracts/src/Staking.sol

```diff
@@ -0,0 +1,476 @@
+        if (validators.length != isRegistration.length) revert ArrayLengthMismatch();
+
+        uint256 executableAt = block.timestamp + configTimeDelay;
+        bytes32 validatorsHash = keccak256(abi.encode(validators, isRegistration, executableAt));
+        for (uint256 i = 0; i < validators.length; i++) {
+            if (validators[i] == address(0)) revert InvalidAddress();
+        }
+
+        pendingValidatorChangeHash = validatorsHash;
+
+        emit ValidatorsProposed(validatorsHash, validators, isRegistration, executableAt);
+    }
+
+    // ============================================================
+    // EXTERNAL FUNCTIONS - CONFIGURATION EXECUTION (PUBLIC)
+    // ============================================================
+
+    /*
+     * @notice Execute a pending withdraw delay change
+     */
+    function executeWithdrawDelayChange() external {
+        ConfigProposal memory proposal = pendingWithdrawDelayChange;
+        if (proposal.executableAt == 0) revert NoProposalExists();
+        if (block.timestamp < proposal.executableAt) revert ProposalNotExecutable();
+
+        uint256 oldDelay = withdrawDelay;
+        withdrawDelay = uint128(proposal.value);
+        delete pendingWithdrawDelayChange;
+        emit WithdrawDelayChanged(oldDelay, proposal.value);
+    }
+
+    /*
+     * @notice Execute pending validator changes
+     */
+    function executeValidatorChanges(
+        address[] calldata validators,
+        bool[] calldata isRegistration,
+        uint256 executableAt
+    ) external {
+        bytes32 proposalHash = pendingValidatorChangeHash;
+        if (proposalHash == bytes32(0)) revert NoProposalExists();
+
+        bytes32 validatorsHash = keccak256(abi.encode(validators, isRegistration, executableAt));
+        if (proposalHash != validatorsHash) revert InvalidProposalHash();
+        if (executableAt == 0) revert ProposalNotSet();
+        if (block.timestamp < executableAt) revert ProposalNotExecutable();
+
+        for (uint256 i = 0; i < validators.length; i++) {
+            isValidator[validators[i]] = isRegistration[i];
+            emit ValidatorUpdated(validators[i], isRegistration[i]);
+        }
+
+        pendingValidatorChangeHash = bytes32(0);
+    }
+
+    // ============================================================
+    // EXTERNAL FUNCTIONS - TOKEN RECOVERY (OWNER ONLY)
+    // ============================================================
+
+    /*
+     * @notice Recover accidentally sent tokens
+     * @param token The token address to recover
+     * @param to The address to send recovered tokens to
+     */
+    function recoverTokens(address token, address to) external onlyOwner {
+        if (to == address(0)) revert InvalidAddress();
+
+        uint256 recoverable;
+        if (token == address(safeToken)) {
+            uint256 balance = safeToken.balanceOf(address(this));
+            recoverable = balance - totalStakedAmount - totalPendingWithdrawals;
```

Comment:
Right now, the staking contract does not guarantee an APY - especially because the owner can just withdraw the extra tokens from the staking contract. We discussed this yesterday, but just leaving a note here, since in case we change that property, this code will have to be adjusted as well.